collector/metadata/kubernetes/pod_watch.go

```diff
@@ -248,22 +240,115 @@ func OnUpdate(objOld interface{}, objNew interface{}) {
 		// Two different versions of the same pod will always have different RVs.
 		return
 	}
-	podUpdateMutex.Lock()
-	// TODO: re-implement the updated logic to reduce computation
-	onDelete(objOld)
+
+	oldCachePod, ok := globalPodInfo.get(oldPod.Namespace, oldPod.Name)
+	if !ok {
+		onAdd(objNew)
+		return
+	}
+	// Always override the old pod in the cache
 	onAdd(objNew)
-	podUpdateMutex.Unlock()
+
+	// Delay delete the pod using the difference between the old pod and the new one
+	deletedPodInfo := &deletedPodInfo{
+		name:         "",
+		namespace:    oldPod.Namespace,
+		containerIds: nil,
+		ip:           oldPod.Status.PodIP,
+		ports:        nil,
+		hostIp:       oldPod.Status.HostIP,
+		hostPorts:    nil,
+	}
+
+	if oldPod.Name != newPod.Name {
+		deletedPodInfo.name = oldPod.Name
+	}
+
+	// Check the containers' ID
+	newContainerIds := make([]string, 0)
+	for _, containerStatus := range oldPod.Status.ContainerStatuses {
+		shortenContainerId := TruncateContainerId(containerStatus.ContainerID)
+		if shortenContainerId == "" {
+			continue
+		}
+		newContainerIds = append(newContainerIds, shortenContainerId)
+	}
```

Comment:
Here we need to search newContainerIds  from newPod.status.ContainerStatues instead of `oldPod`

Reply:
Sorry, it's a typo.